Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -44,6 +44,115 @@ static wstring __GetUserDefaultLocaleName() {
+    CFMutableStringRef ret = CFStringCreateMutableCopy(nullptr, CFStringGetLength(string), string);
+    CFStringUppercase(ret, locale);
+    return ret;
+}
+
+/**
+ * Private helper that creates the IDWriteFontList for a given font family name
+ */
+HRESULT __DWriteGetFontListForFamilyName(CFStringRef familyName, IDWriteFontList** outFontList) {
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    // Get the system font collection.
+    ComPtr<IDWriteFontCollection> fontCollection;
+    RETURN_IF_FAILED(dwriteFactory->GetSystemFontCollection(&fontCollection));
+
+    // Get the font family.
+    CFIndex familyNameLength = CFStringGetLength(familyName);
+    std::vector<UniChar> unicharFamilyName(familyNameLength + 1);
+    CFStringGetCharacters(familyName, CFRangeMake(0, familyNameLength), unicharFamilyName.data());
+
+    size_t index = 0;
+    BOOL exists = false;
+
+    RETURN_IF_FAILED(fontCollection->FindFamilyName(reinterpret_cast<wchar_t*>(unicharFamilyName.data()), &index, &exists));
+    RETURN_HR_IF(E_INVALIDARG, !exists);
+
+    ComPtr<IDWriteFontFamily> fontFamily;
+    RETURN_IF_FAILED(fontCollection->GetFontFamily(index, &fontFamily));
+
+    RETURN_IF_FAILED(
+        fontFamily->GetMatchingFonts(DWRITE_FONT_WEIGHT_NORMAL, DWRITE_FONT_STRETCH_NORMAL, DWRITE_FONT_STYLE_NORMAL, outFontList));
+
+    return S_OK;
+}
+
+/**
+ * Private struct that exposes CFHash and CFEqual to an STL container
+ */
+struct __CFStringHashEqual {
+    std::size_t operator()(const woc::unique_cf<CFStringRef>& item) const {
+        return CFHash(item.get());
+    }
+
+    bool operator()(const woc::unique_cf<CFStringRef>& item1, const woc::unique_cf<CFStringRef>& item2) const {
+        return CFEqual(item1.get(), item2.get());
+    }
+};
+
+/**
+ * Private static map, that maps uppercase display name and postscript name, to a _DWriteFontProperties struct for the corresponding font.
+ */
+static std::unordered_map<woc::unique_cf<CFStringRef>, std::shared_ptr<_DWriteFontProperties>, __CFStringHashEqual, __CFStringHashEqual>
+    c_fontPropertiesMap = []() {
+        decltype(c_fontPropertiesMap) ret;
+
+        woc::unique_cf<CFArrayRef> familyNames(_DWriteCopyFontFamilyNames());
+        woc::unique_cf<CFLocaleRef> locale(CFLocaleCopyCurrent());
+
+        for (size_t i = 0; i < CFArrayGetCount(familyNames.get()); ++i) {
+            woc::unique_cf<CFStringRef> familyName(static_cast<CFStringRef>(CFRetain(CFArrayGetValueAtIndex(familyNames.get(), i))));
+            ComPtr<IDWriteFontList> fontList;
+
+            if (FAILED(__DWriteGetFontListForFamilyName(familyName.get(), &fontList))) {
+                continue;
+            }
+
+            size_t count = fontList->GetFontCount();
+
+            for (size_t j = 0; j < count; j++) {
+                ComPtr<IDWriteFont> font;
+                if (FAILED(fontList->GetFont(j, &font))) {
```

Comment:
this version is most likely (can't say for sure) more performant than our previous version, as it lacks the string parsing. we can make this a little more lazy, but not by much - we can't avoid iterating through all fonts and families. we can do it later, but it would probably hurt readability (see other comment) for now, leaving as-is. we can pursue optimizations at another point.

---
In reply to: [87664127](https://github.com/Microsoft/WinObjC/pull/1366#discussion_r87664127) [](ancestors = 87664127)